src/Runtime/Factors/StringFormatOp.cs

```diff
@@ -754,15 +756,38 @@ public static double LogEvidenceRatio(string str, StringDistribution format, ILi
         /// <returns>The generated array of argument names.</returns>
         private static string[] GetArgumentNames(int argCount)
         {
-            string[] result;
-            if (!ArgumentCountToNames.TryGetValue(argCount, out result))
+            string[] result = null;
+            var cache = GetCache(argCount);
+            var readOnlyResult = cache[argCount];
+            if (readOnlyResult == null)
             {
                 result = Util.ArrayInit(argCount, i => i.ToString(CultureInfo.InvariantCulture));
-                ArgumentCountToNames.Add(argCount, result);
+                ArgumentCountToNames[argCount] = result;
+            }
+            else
+            {
+                result = new List<string>(readOnlyResult).ToArray();
             }
 
             return result;
         }
+
+        private static ReadOnlyArray<string>?[] GetCache(int minSize)
+        {
+            lock (ArgumentCountToNames)
+            {
+                if (minSize >= ArgumentCountToNames.Length)
+                {
+                    var oldCache = ArgumentCountToNames;
+                    var newCache = new ReadOnlyArray<string>?[minSize + 1];
```

Comment:
set the size to 2 * Min(minSize, ArgumentCountToNames).
Exponential growing strategy is always better